standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/StatisticsManagementTask.java

```diff
@@ -0,0 +1,162 @@
+ *
+ */
+public class StatisticsManagementTask implements MetastoreTaskThread {
+    private static final Logger LOG = LoggerFactory.getLogger(StatisticsManagementTask.class);
+
+    // global
+    public static final String STATISTICS_AUTO_DELETION = "statistics.auto.deletion";
+    public static final String STATISTICS_RETENTION_PERIOD = "statistics.retention.period";
+
+    // The 2 configs for users to set in the conf
+    // this is an optional table property, if this property does not exist for a table, then it is not excluded
+    public static final String STATISTICS_AUTO_DELETION_EXCLUDE_TBLPROPERTY = "statistics.auto.deletion.exclude";
+
+    private static final Lock lock = new ReentrantLock();
+
+    // these are just for testing
+    private static int completedAttempts;
+    private static int skippedAttempts;
+
+    private Configuration conf;
+
+    @Override
+    public long runFrequency(TimeUnit unit) {
+        return MetastoreConf.getTimeVar(conf, MetastoreConf.ConfVars.STATISTICS_MANAGEMENT_TASK_FREQUENCY, unit);
+    }
+
+    @Override
+    public void setConf(Configuration configuration) {
+        // we modify conf in setupConf(), so we make a copy
+        conf = new Configuration(configuration);
+    }
+
+    @Override
+    public Configuration getConf() {
+        return conf;
+    }
+
+    // what needs to be included in this run() method:
+    // get the "lastAnalyzed" information from TAB_COL_STATS and find all the tables need to be deleted
+    // delete all column stats
+    @Override
+    public void run() {
+        if (LOG.isDebugEnabled()) {
+            LOG.debug("Auto statistics deletion started. Cleaning up table/partition column statistics over the retention period.");
+        }
+        long retentionMillis = MetastoreConf.getTimeVar(conf, MetastoreConf.ConfVars. STATISTICS_RETENTION_PERIOD, TimeUnit.MILLISECONDS);
+        if (retentionMillis <= 0 || !MetastoreConf.getBoolVar(conf, MetastoreConf.ConfVars.STATISTICS_AUTO_DELETION)) {
+            LOG.info("Statistics auto deletion is set to off currently.");
+            return;
+        }
+        if (lock.tryLock()) {
+            skippedAttempts = 0;
+            String qualifiedTableName = null;
+            IMetaStoreClient msc = null;
+            try {
+                // Get retention period in conf in milliseconds; default is 365 days.
+                long now = System.currentTimeMillis();
+                long lastAnalyzedThreshold = (now - retentionMillis) / 1000;
+
+                // Get all databases from metastore
+                List<String> databases = msc.getAllDatabases();
+                RawStore ms = HMSHandler.getMSForConf(conf);
+                PersistenceManager pm = ((ObjectStore) ms).getPersistenceManager();
+                Query q = pm.newQuery("SELECT FROM org.apache.hadoop.hive.metastore.model.MTableColumnStatistics");
+                q.setFilter("lastAnalyzed < " + lastAnalyzedThreshold);
```

Comment:
Why don't we do something like this:
q.setFilter(filter.toString());
q.declareParameters(paramStr.toString());